File: common-problems-leetcode/group-anagrams.js
/*
Given an array of strings, group anagrams together.

Example:

Input: ["eat", "tea", "tan", "ate", "nat", "bat"],
Output:
[
  ["ate","eat","tea"],
  ["nat","tan"],
  ["bat"]
]
Note:

All inputs will be in lowercase.
The order of your output does not matter.
Accepted
323,183
Submissions
700,745
*/
// 5th percentile. 5136ms
// Amazingly bad solution
/**
 * @param {string[]} strs
 * @return {string[][]}
 */
var groupAnagrams = function(strs) {
  let groups = {};
  
  strs.forEach(s => {
      key = s.split('').sort().join('');
      if (Object.keys(groups).indexOf(key) === -1) {
          groups[key] = [s];
      } else{
          groups[key].push(s);
      }
  })
  
  let res = [];
  Object.keys(groups).forEach(group => {
      res.push(groups[group]);
  })
  return res;
};



// 94th percentile. 124 ms.
// Maps are once again the best thing ever
/**
 * @param {string[]} strs
 * @return {string[][]}
 */
var groupAnagrams = function(strs) {
  let groups = new Map()
  
  strs.forEach(s => {
      key = s.split('').sort().join('');
      if ( !groups.has(key) ) {
          groups.set(key, [s]);
      } else{
          groups.get(key).push(s)
      }
  })
  
  let res = [];
  groups.forEach(group => {
      res.push(group);
  })
  return res;
};

File: common-problems-leetcode/group-anagrams.java
/*
Given an array of strings, group anagrams together.

Example:

Input: ["eat", "tea", "tan", "ate", "nat", "bat"],
Output:
[
  ["ate","eat","tea"],
  ["nat","tan"],
  ["bat"]
]
Note:

All inputs will be in lowercase.
The order of your output does not matter.
Accepted
323,183
Submissions
700,745
*/
// 13th percentile. 41ms. 
import java.util.Map;
import java.util.HashMap;
import java.util.List;
import java.util.ArrayList;
import java.util.Arrays;
class Solution {
  public List<List<String>> groupAnagrams(String[] strs) {
      HashMap<String, List> groups = new HashMap<String, List>();
      
      for (String s : strs) {
          char[] cs = s.toCharArray();
          Arrays.sort(cs);
          String key = new String(cs);
          if (!groups.containsKey(key)) {
              ArrayList<String> temp = new ArrayList<String>(); 
              temp.add(s);
              groups.put(key, temp);
          } else {
              List<String> temp = groups.get(key);
              temp.add(s);
              groups.put(key, temp);
          }
      }
      
      List<List<String>> res = new ArrayList<List<String>>();
      groups.forEach( (k,v) ->
                    res.add(v));
      
      return res;
  }
}



// 15th percentile 37ms.
// Just clarified generics
class Solution {
  public List<List<String>> groupAnagrams(String[] strs) {
      HashMap<String, List<String>> groups = new HashMap<String, List<String>>();
      
      for (String s : strs) {
          char[] cs = s.toCharArray();
          Arrays.sort(cs);
          String key = new String(cs);
          if (!groups.containsKey(key)) {
              ArrayList<String> temp = new ArrayList<String>(); 
              temp.add(s);
              groups.put(key, temp);
          } else {
              List<String> temp = groups.get(key);
              temp.add(s);
              groups.put(key, temp);
          }
      }
      
      List<List<String>> res = new ArrayList<List<String>>();
      groups.forEach( (k,v) ->
                    res.add(v));
      
      return res;
  }
}

// 92nd percentile. 9ms
// So much faster with traditional for each loop
class Solution {
  public List<List<String>> groupAnagrams(String[] strs) {
      HashMap<String, List<String>> groups = new HashMap<String, List<String>>();
      
      for (String s : strs) {
          char[] cs = s.toCharArray();
          Arrays.sort(cs);
          String key = new String(cs);
          if (!groups.containsKey(key)) {
              ArrayList<String> temp = new ArrayList<String>(); 
              temp.add(s);
              groups.put(key, temp);
          } else {
              List<String> temp = groups.get(key);
              temp.add(s);
              groups.put(key, temp);
          }
      }
      
      List<List<String>> res = new ArrayList<List<String>>();
      for (String s : groups.keySet()) {
          res.add(groups.get(s));
      }
      
      return res;
  }
}